### ISSUE TITLE
Container creation fails for images with too many layers

### ISSUE THREAD

**Issue by imre-kerr-sb1** [nestybox/sysbox#634]
## Description
When running a container with too many layers (71 or more in my testing), docker fails with the following message:

>docker: Error response from daemon: failed to create shim task: OCI runtime create failed: error in the container spec: failed to request rootfs cloning from sysbox-mgr: failed to invoke ReqCloneRootfs via grpc: rpc error: code = Unknown desc = failed to mount clone for container 99e8e3a5d6a5: failed to set up bottom ovfs mount: failed to mount overlayfs on /var/lib/sysbox/rootfs/99e8e3a5d6a5ca47390327b57b581fbfeb76dcbb21c8ef1de703a1a0e1c51e73/bottom/merged: no such file or directory: unknown.
>ERRO[0000] error waiting for container: context canceled 

## Steps to reproduce
1. Create a Dockerfile with the following content:
```Dockerfile
From busybox:latest

RUN mktemp
RUN mktemp
# ... Repeat 100 times
```
2. `docker build . -t too-many-layers`
3.  `docker run --runtime=sysbox-runc too-many-layers`

## System info

* Sysbox version: 0.5.2
* Docker version: 20.10.23, build 7155243
* Linux version: 5.15.0-1026-aws
* Distribution: Ubuntu 22.04.1 LTS

**Comment by ctalledo** [comment]
Thanks @imre-kerr-sb1 for reporting the error; have never seen this one before.

Curious how you hit it? Did you come across it with a real image?

**Comment by imre-kerr-sb1** [comment]
>Curious how you hit it? Did you come across it with a real image?

Yes. The image in question was based on `gitpod/workspace-full` (54 layers), with 26 layers of our own stuff on top. Of course this is easily fixable by being more economical with layers or even squashing them. But I still figured I should report it as it's obviously a bug.

**Comment by ctalledo** [comment]
Hi @commiterate, are you hitting the exact same error in the original description?

What kernel version do you have? 

I ask because the error in the original description should never occur with kernel 5.19+ which supports a kernel feature called `idmapped` mounts. Even with kernels < 5.19, it should not occur if Sysbox is using the [shiftfs](https://github.com/nestybox/sysbox/blob/master/docs/user-guide/design.md#shiftfs-module) kernel module.

**Comment by ctalledo** [comment]
Thanks @commiterate ; can you paste the output of the `journalctl -u sysbox-mgr`? I am interested in the first 15 or so lines it emits after starting, where it checks if `shiftfs` is working properly on the host.

**Comment by ctalledo** [comment]
> Here's the sysbox-mgr startup logs (did 0.6.5 first before downgrading back to 0.6.4).
> It looks like 0.6.4 thinks shiftfs works properly but 0.6.5 doesn't.

Thanks that helps. Yes that would explain why downgrading from v0.6.5 to v0.6.4 works.

So there's two issues here:

1) In v0.6.5, the shiftfs checker in sysbox-mgr fails, yet with v0.6.4 it passes. Not sure why, need to check what changed.

2) Even without shiftfs, sysbox-mgr falls-back to a technique called "rootfs cloning". For some reason that technique is failing when an image has many layers apparently. 

Regarding issue (1), it must be one of these two commits:

https://github.com/nestybox/sysbox-libs/commit/1e5a7ea2d4fcc86e8021e6d82ad31caacfeb2116
https://github.com/nestybox/sysbox-libs/commit/33cbdf5a9e98447a8142f0365922ab1608afaaa1

I can't reproduce it however.

One way to learn more is to re-start `sysbox-mgr` with the `--log-level=debug` flag, by editing the sysbox-mgr systemd service, as described [here](https://github.com/nestybox/sysbox/blob/master/docs/user-guide/configuration.md#reconfiguration-procedure). That will output more info during the shiftfs check.

However, the best way to overcome this whole issue is for you to upgrade the AWS node to kernel 5.19 or later. The reason is that this kernel supports a feature called "overlayfs on idmapped mounts", which is the in-kernel official replacement for shiftfs. Sysbox will automatically detect this and use this kernel feature, which will make the problem go away.

**Comment by ctalledo** [comment]
Thanks that helps.

> That's a bit surprising given that all of the Sysbox systemd units are running as root

Yes correct, but that code that is performing the shiftfs check creates a child process that runs in a Linux user-namespace, thereby is less privileged. 

However, sysbox-mgr [temporarily opens the permissions](https://github.com/nestybox/sysbox-mgr/blob/2a8aff4bc6bb16945cb49ea515bd45d1be10f8cf/utils.go#L960) under `/var/lib/sysbox` during the check for precisely this same reason.

But I just noticed it chmods `/var/lib/sysbox` to [0755](https://github.com/nestybox/sysbox-mgr/blob/2a8aff4bc6bb16945cb49ea515bd45d1be10f8cf/utils.go#L969), which is still to restrictive since the shiftfs checker needs to create a file and therefore needs "other-write" permissions (e.g., 0777 would have been better).

**Comment by ctalledo** [comment]
BTW, the reason v0.6.4 works is that the shiftfs checker was not as complete as its in v0.6.5. That is, the v0.6.5 shiftfs checker is better, but it seems it needs more permissions that it currently has. 

I'll open a PR so that Sysbox v0.6.6 will carry the fix.

**Comment by ctalledo** [comment]
Fix PR: https://github.com/nestybox/sysbox-mgr/pull/73
